core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -265,17 +283,51 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
             return allocationService.applyFailedShards(currentState, failedShards);
         }
 
-        private boolean shardExists(ClusterState currentState, ShardRoutingEntry task) {
+        private enum ValidationResult {
+            VALID,
+            SOURCE_INVALID,
+            SHARD_MISSING
+        }
+
+        private ValidationResult validateTask(ClusterState currentState, ShardRoutingEntry task) {
+
+            // non-local requests
+            if (!task.shardRouting.isSameAllocation(task.sourceShardRouting)) {
+                IndexRoutingTable indexRoutingTable = currentState.getRoutingTable().index(task.getShardRouting().index().getName());
```

Comment:
That's the second time I see this pattern. Can you we add an shardRoutingTableOrNull utility method to RoutingTable?